compose/config/environment.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import logging
+import os
+
+from .errors import ConfigurationError
+
+log = logging.getLogger(__name__)
+
+
+class BlankDefaultDict(dict):
+    def __init__(self, *args, **kwargs):
+        super(BlankDefaultDict, self).__init__(*args, **kwargs)
+        self.missing_keys = []
+
+    def __getitem__(self, key):
+        try:
+            return super(BlankDefaultDict, self).__getitem__(key)
+        except KeyError:
+            if key not in self.missing_keys:
+                log.warn(
+                    "The {} variable is not set. Defaulting to a blank string."
+                    .format(key)
+                )
+                self.missing_keys.append(key)
+
+            return ""
+
+
+class Environment(BlankDefaultDict):
+    __instance = None
+
+    @classmethod
+    def get_instance(cls, base_dir='.'):
```

Comment:
The first version of this method had base_dir be `None` by default, and fail if it was called for the first time with no value. However, this caused errors with our test suite which does not necessarily calls the code through expected paths.

Current directory seems to be a sane default for 99.9% of cases.